### DISCUSSION TITLE
Refresh cache after `revalidateTag`

### DISCUSSION THREAD

**Discussion by LasseRosenow** [vercel/next.js#65860] [Help]
### Summary

I have an api route for on demand revalidation using `revalidateTag`.
This works. But as I see is that this only purges the cache but does not refetch it.
Refetch happens on page visit.

But what I need is to actually refetch all the purged cache immediately.
For my rather small page this is the perfect solution that would give my users the fastest experience.

Is there an official way to do this?

### Additional information

```js
I use next.js 14 app router
```

**Comment by icyJoseph** [comment]
```js
router.refresh()
```

**Reply by LasseRosenow** [reply]
> ```js
> router.refresh()
> ```

But that is only for the current route isn't it?
I need to refresh the cache of all routes, or more specifically of all fetches that had that tag.

**Reply by icyJoseph** [reply]
But, maybe I rushed reading, but the issue you have is that the client side cache is not revalidated right? Let me check some resources

**Reply by icyJoseph** [reply]
Could you read the router refresh section on this deep dive? Doesn't that sound like what you need?

**Reply by LasseRosenow** [reply]
First of all: Thank you for taking your time, trying to help me! :)

Actually it is not about the client side at all.

I have an API route under `/api/revalidate` and in that route handler I call `revalidateTag("all")`.
(This endpoint gets called by my headless CMS, whenever some content changes somewhere in my website.)
Every fetch request of my website uses the "all" tag.
So the server removes all the cache of these fetches from the CMS.
But the problem is, that the server does not update them.
So when I now visit my website, the server will update the caches for those server side queries just at that time, which increases the load time of the website.

I want the server to update all the server-side fetches after i invalidated them.
So my users will immediately get the prerendered sites and they don't just get rendered on request.

Right now I am working around this by just calling a fetch() request on every possible route, after I revalidated it, but that is really hacky and also needs to be adjusted to each project. A more general solution would be great :)

**Comment by J4v4Scr1pt** [comment]
I have a simillar issue,I'm trying to understand how revalidateTag work. I'm used to Tanstack, where you from anywhere in the application can invalidate a cache and tanstack will refresh data associated with that tag. I'm using revalidateTag in the same manor and it works locally as I thought. But when I deploy the application it does not work as I think, it returns the initalComponent no matter what, with a hard-refresh it flickers from correct to wrong. I need to open a new tab and visit the site to get the correct component based on the data. 

I have a serveraction that is being called after a user saves some answer, what I want is to update a component inside of my header that check if the user have answered the question.

Where I intally get the data to the Header(ServerSide):

```
const [configuration, customer, user] = await Promise.all([
        GET<Configuration>('GetConfiguration'),
        GET<Customer>('GetCustomer'),
        GET<User>('GetUser', 'User'),
    ]);
```

Then I pass the data to a component:

```
<HeaderRightContent
                canAnswerQuestion={user?.canAnswerQuestion || false}
                customerName={customer?.Name}>
```

Consuming component (Simplyfied):

```
export function AnswerButton({ canAnswerQuestion}: { canAnswerQuestion: boolean }) { 
return (
                {canAnswerDailyQuestion ? (
                    <Button onPress={() => openModal}>
                        Answer
                    </Button>
                ) : (<div>Done<div>)}
}
```

ServerAction:

```
'use server';

import { POST } from '@/utils/api';
import { Answer } from '../types';
import { revalidateTag } from 'next/cache';

export async function postAnswer(data: Answer) {
    let success = false;
    try {
        const response = await POST('SaveAnswer', JSON.stringify(data));
        success = response.Success;
        revalidateTag('User');
        return success;
    } catch {}
    return success;
}
```

fetch util. function:

```
export async function GET<T>(
    endpoint: string,
    cacheTag?: string,
    cache: RequestCache | undefined = 'force-cache'
): Promise<T | null | undefined> {
    const neededCookies = cookies().toString();

    const res = await fetch(`${process.env.PATH_INTERNAL}${endpoint}`, {
        method: 'GET',
        cache,
        headers: {
            'Content-Type': 'application/json',
            Cookie: neededCookies || '',
        },
        redirect: 'manual',
        ...(cacheTag && { next: { tags: [cacheTag] } }),
    });

    if (res.status >= 300 && res.status < 400) permanentRedirect(redirectLoginUrl);

    try {
        return await res.json();
    } catch (err) {
        return null;
    }
}
export async function POST(endpoint: string, data?: string) {
    const neededCookies = cookies().toString();
    const res = await fetch(`${process.env.PATH_INTERNAL}${endpoint}`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            Cookie: neededCookies || '',
        },
        body: data,
        redirect: 'manual',
    });

    if (res.status >= 300 && res.status < 400) permanentRedirect(redirectLoginUrl);

    try {
        return await res.json();
    } catch (err) {
        return null;
    }
}
```

What I would expect is nextjs to refecth data related to the tag and re-render affected components. But it does not. What I'm doing wrong here?
